api/auth.proto

```diff
@@ -44,3 +46,73 @@ message AuthAction {
   // List of rules
   repeated Rule rule = 2;
 }
+
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // Address of the downstream source of the request.
+  Address source = 1;
+
+  // Address of the upstream destination of the request.
+  Address destination = 2; 
+
+  // If the request is an HTTP request, the parameters
+  // of the HTTP request.
+  HTTPRequest http = 3;
+
+  // If the request came over a secure TLS connection, information about the connection and 
+  // downstream client.
+  TLSConnection tls = 4;
+}
+
+message HTTPRequest {
+  // HTTP Version of the request, e.g. "1.1"
+  string version = 1;
+
+  // Method, e.g. "GET"
+  string method = 2;
+
+  // The value of the Host: header
+  string host = 3;
+
+  // The request URI path.
+  string path = 4;
+
+  // Additional HTTP headers and their values.
+  map<string, string> headers = 5;
+}
+
+message TLSConnection {
+  // True if the connection is encrypted with TLS.  False otherwise. 
+  bool tls_encrypted = 1;
+
+  // If the client presented an X.509 certificate, it is contained here.
+  X509Cert client_cert = 2;
+
+  // The server certificate used to secure the connection.
+  X509Cert server_cert = 3;
+}
+
+
+// An X.509 certificate, with any corresponding fields already decoded by Envoy included.
+message X509Cert {
+  // The raw certificate as a byte array.
+  bytes raw = 1;
+
+  // The decoded Subject string.
+  string subject = 2;
+
+  // A SHA256-hash of the certificate.
+  string sha_256_hash = 3;
+
+  // Decoded Subject Alternative Names on the certificate. 
+  repeated string subject_alt_names = 4;
+}
+
+
+message CheckResponse {
+  // Return OK to allow the request.  Any other response indicates the request should be denied.
+  google.rpc.Status status = 1;
```

Comment:
Caching protocol that returns information that was used in decision making.
Client forms a cache key based on the response.

```
message Referenced {   
   // True if the source field was used in making the decision.
   bool source = 1;

   // True if the destination field was used in making the decision.
   bool destination = 2;

   // HTTPRequest.version / method / host should always be part of the cache key
   // so they are omitted from there.
   
   // path is highly variable so regex match is needed.
   // If nil, path was not used to make the decision.
   Match path = 3;

  // The following headers were used to make a decision.
   repeated Match headers = 4;
}
 
message Match {
    // The kind of match against the attribute value.
    Condition condition = 1;

    // The matching regex in the case of condition = REGEX
    string regex = 2;
   
   // match a specific key in the string map
   string header = 4;
  }

 // How an  value was matched
  enum Condition {
    CONDITION_UNSPECIFIED = 0;    // should not occur
    ABSENCE = 1;                  // match when attribute doesn't exist
    EXACT = 2;                    // match when attribute value is an exact byte-for-byte match
    REGEX = 3;                    // match when attribute value matches the included regex
  }

```

Reply:
And similar to above for referenced attributes

```
// Describes the attributes that were used to determine the response.
// This can be used to construct a response cache.
message ReferencedAttributes {
  message AttributeMatch {
    // The name of the attribute. 
    string name = 1;

    // The kind of match against the attribute value.
    Condition condition = 2;

    // The matching regex in the case of condition = REGEX
    string regex = 3;
   
   // match a specific key in the string map
   string map_key = 4;
  }
  // Describes a set of attributes.
  repeated AttributeMatch attribute_matches = 2 ;
}
```

Reply:
I think the cache proposal is underbaked compared to what Istio is doing in Mixer (for example, there is no protocol for cache validity), so Mixer won't be able to use it without additional work.

Speaking for my team, I don't think we want to commit to building a cache in the first iteration of the external auth filter.

I propose we add caching sufficient for Mixer use cases when we converge the proposals.  Until then I think it would just languish unimplemented.